<a href="https://colab.research.google.com/github/duybluemind1988/Data-science/blob/master/Recommendation/Kaggle_movie_recommender_all_methods.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#!pip install surprise

     |████████████████████████████████| 11.8MB 4.5MB/s 
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.1-cp36-cp36m-linux_x86_64.whl size=1670921 sha256=e5e6e69c9774b6182b90d8d9d39a28da25a6e546ea2455f4f1b168b9f3311d53
  Stored in directory: /root/.cache/pip/wheels/78/9c/3d/41b419c9d2aff5b6e2b4c0fc8d25c538202834058f9ed110d0
Successfully built scikit-surprise


In [2]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from nltk.corpus import wordnet
#from surprise import Reader, Dataset, SVD
#from surprise.model_selection import cross_validate #replace for evaluate()

import warnings; warnings.simplefilter('ignore')

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [3]:
movies_metadata=pd.read_csv('/content/drive/My Drive/Data/Movie dataset/movies_metadata.csv',encoding = "ISO-8859-1")
print(movies_metadata.shape)
movies_metadata.head(2)

(45466, 24)


adult  ... vote_count
0  False  ...     5415.0
1  False  ...     2413.0

[2 rows x 24 columns]

In [4]:
movies=pd.read_csv('/content/drive/My Drive/Data/Movie dataset/Small/movies.csv',encoding = "ISO-8859-1")
print(movies.shape)
movies.head()

(2500, 3)


movieId  ...                                       genres
0        1  ...  Adventure|Animation|Children|Comedy|Fantasy
1        2  ...                   Adventure|Children|Fantasy
2        3  ...                               Comedy|Romance
3        4  ...                         Comedy|Drama|Romance
4        5  ...                                       Comedy

[5 rows x 3 columns]

In [5]:
ratings=pd.read_csv('/content/drive/My Drive/Data/Movie dataset/Small/ratings.csv',encoding = "ISO-8859-1")
print(ratings.shape)
ratings.head()

(264505, 4)


userId  movieId  rating   timestamp
0   12882        1     4.0  1147195252
1   12882       32     3.5  1147195307
2   12882       47     5.0  1147195343
3   12882       50     5.0  1147185499
4   12882      110     4.5  1147195239

In [6]:
tags=pd.read_csv('/content/drive/My Drive/Data/Movie dataset/Small/tags.csv',encoding = "ISO-8859-1")
print(tags.shape)
tags.head()

(94875, 4)


movieId  userId           tag   timestamp
0     3916   12882        sports  1147195545
1     4085   12882  Eddie Murphy  1147195966
2    33660   12882        boxing  1147195514
3     1197     320     must show  1145964801
4     1396     320     must show  1145964810

# Simple recommender

## Method 1: suggestion only by score

In [ ]:
df=movies_metadata.copy()
df.head(2)

adult  ... vote_count
0  False  ...     5415.0
1  False  ...     2413.0

[2 rows x 24 columns]

In [ ]:
#Calculate the number of votes garnered by the 80th percentile movie
m = df['vote_count'].quantile(0.80)
m

50.0

In [ ]:
#Only consider movies longer than 45 minutes and shorter than 300 minutes
q_movies = df[(df['runtime'] >= 45) & (df['runtime'] <= 300)]

#Only consider movies that have garnered more than m votes
q_movies = q_movies[q_movies['vote_count'] >= m]

#Inspect the number of movies that made the cut
print(q_movies.shape)
q_movies.head(2)

(8963, 24)


adult  ... vote_count
0  False  ...     5415.0
1  False  ...     2413.0

[2 rows x 24 columns]

In [ ]:
# Calculate C
C = df['vote_average'].mean()
C

5.618207215133889

In [ ]:
# Function to compute the IMDB weighted rating for each movie
# m: number of vote average (50)
# C: vote average (5.618) 
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    # Compute the weighted score
    return (v/(v+m) * R) + (m/(m+v) * C)

In [ ]:
# Compute the score using the weighted_rating function defined above
q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

#Sort movies in descending order of their scores
q_movies = q_movies.sort_values('score', ascending=False)

#Print the top 25 movies
q_movies[['title', 'vote_count', 'vote_average', 'score', 'runtime']].head(25)

title  vote_count  ...     score  runtime
10309      Dilwale Dulhania Le Jayenge       661.0  ...  8.855148    190.0
314           The Shawshank Redemption      8358.0  ...  8.482863    142.0
834                      The Godfather      6024.0  ...  8.476278    175.0
40251                       Your Name.      1030.0  ...  8.366584    106.0
12481                  The Dark Knight     12269.0  ...  8.289115    152.0
2843                        Fight Club      9678.0  ...  8.286216    139.0
292                       Pulp Fiction      8670.0  ...  8.284623    154.0
522                   Schindler's List      4436.0  ...  8.270109    195.0
23673                         Whiplash      4376.0  ...  8.269704    105.0
5481                     Spirited Away      3968.0  ...  8.266628    125.0
2211                 Life Is Beautiful      3643.0  ...  8.263691    116.0
1178            The Godfather: Part II      3418.0  ...  8.261335    200.0
1152   One Flew Over the Cuckoo's Nest      3001.0  ...  8.256051    133.0
1176                            Psycho      2405.0  ...  8.245381    109.0
351                       Forrest Gump      8147.0  ...  8.184252    142.0
1184       Once Upon a Time in America      1104.0  ...  8.183804    229.0
1154           The Empire Strikes Back      5998.0  ...  8.178656    124.0
18465                 The Intouchables      5410.0  ...  8.176357    112.0
289             Leon: The Professional      4293.0  ...  8.170276    110.0
3030                    The Green Mile      4166.0  ...  8.169381    189.0
1170                        GoodFellas      3211.0  ...  8.160414    145.0
2216                American History X      3120.0  ...  8.159278    119.0
1161                      12 Angry Men      2130.0  ...  8.140785     96.0
9698              Howl's Moving Castle      2049.0  ...  8.138499    119.0
2884                 Princess Mononoke      2041.0  ...  8.138264    134.0

[25 rows x 5 columns]

## Method 2 (best) suggestion by genres and score

In [ ]:
md=movies_metadata.copy()
md.head(2)

adult  ... vote_count
0  False  ...     5415.0
1  False  ...     2413.0

[2 rows x 24 columns]

In [ ]:
md['genres'] = md['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

In [ ]:
print(md.shape)

(45466, 24)


I use the TMDB Ratings to come up with our **Top Movies Chart.** I will use IMDB's *weighted rating* formula to construct my chart. Mathematically, it is represented as follows:

Weighted Rating (WR) = $(\frac{v}{v + m} . R) + (\frac{m}{v + m} . C)$

where,
* *v* is the number of votes for the movie
* *m* is the minimum votes required to be listed in the chart
* *R* is the average rating of the movie
* *C* is the mean vote across the whole report

The next step is to determine an appropriate value for *m*, the minimum votes required to be listed in the chart. We will use **95th percentile** as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 95% of the movies in the list.

I will build our overall Top 250 Chart and will define a function to build charts for a particular genre. Let's begin!

In [ ]:
vote_counts = md[md['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = md[md['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
C

5.244896612406511

In [ ]:
m = vote_counts.quantile(0.95)
m

434.0

In [ ]:
md['year'] = pd.to_datetime(md['release_date'], errors='coerce').\
apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)
md['year'] 

0        1995
1        1995
2        1995
3        1995
4        1995
         ... 
45461     NaT
45462    2011
45463    2003
45464    1917
45465    2017
Name: year, Length: 45466, dtype: object

In [ ]:
qualified = md[(md['vote_count'] >= m) & (md['vote_count'].notnull()) & (md['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
print(qualified.shape)
qualified.head()

(2274, 6)


title  year  ...  popularity                            genres
0   Toy Story  1995  ...     21.9469       [Animation, Comedy, Family]
1     Jumanji  1995  ...     17.0155      [Adventure, Fantasy, Family]
5        Heat  1995  ...     17.9249  [Action, Crime, Drama, Thriller]
9   GoldenEye  1995  ...      14.686     [Adventure, Action, Thriller]
15     Casino  1995  ...     10.1374                    [Drama, Crime]

[5 rows x 6 columns]

Filter top movie by weight score

In [ ]:
def weighted_rating(x):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [ ]:
qualified['wr'] = qualified.apply(weighted_rating, axis=1)
qualified = qualified.sort_values('wr', ascending=False).head(250)

In [ ]:
qualified.head(15)

title  ...        wr
15480                                          Inception  ...  7.917588
12481                                    The Dark Knight  ...  7.905871
22879                                       Interstellar  ...  7.897107
2843                                          Fight Club  ...  7.881753
4863   The Lord of the Rings: The Fellowship of the Ring  ...  7.871787
292                                         Pulp Fiction  ...  7.868660
314                             The Shawshank Redemption  ...  7.864000
7000       The Lord of the Rings: The Return of the King  ...  7.861927
351                                         Forrest Gump  ...  7.860656
5814               The Lord of the Rings: The Two Towers  ...  7.851924
256                                            Star Wars  ...  7.834205
1225                                  Back to the Future  ...  7.820813
834                                        The Godfather  ...  7.814847
1154                             The Empire Strikes Back  ...  7.814099
46                                                 Se7en  ...  7.811669

[15 rows x 7 columns]

Filter top movie by genres

In [ ]:
s = md.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'genre'
gen_md = md.drop('genres', axis=1).join(s)

In [ ]:
print(md.shape)
print(gen_md.shape) # rat nhieu film se duoc nhan them nhieu dong do co nhieu the loai

(45466, 25)
(93548, 25)


In [ ]:
gen_md.head()

(93548, 25)


adult                              belongs_to_collection  ...  year      genre
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  1995  Animation
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  1995     Comedy
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  1995     Family
1  False                                                NaN  ...  1995  Adventure
1  False                                                NaN  ...  1995    Fantasy

[5 rows x 25 columns]

In [ ]:
def build_chart(genre, percentile=0.85):
    df = gen_md[gen_md['genre'] == genre]
    vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(percentile)
    
    qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    
    qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(250)
    
    return qualified

In [ ]:
gen_md.genre.value_counts()

Drama                                    20265
Comedy                                   13182
Thriller                                  7624
Romance                                   6735
Action                                    6596
Horror                                    4673
Crime                                     4307
Documentary                               3932
Adventure                                 3496
Science Fiction                           3049
Family                                    2770
Mystery                                   2467
Fantasy                                   2313
Animation                                 1935
Foreign                                   1622
Music                                     1598
History                                   1398
War                                       1323
Western                                   1042
TV Movie                                   767
Vision View Entertainment                    1
BROSTA TV    

In [ ]:
build_chart('Romance').head(10)

title  year  ...  popularity        wr
10309  Dilwale Dulhania Le Jayenge  1995  ...      34.457  8.565285
351                   Forrest Gump  1994  ...     48.3072  7.971357
876                        Vertigo  1958  ...     18.2082  7.811667
40251                   Your Name.  2016  ...   34.461252  7.789489
883               Some Like It Hot  1959  ...     11.8451  7.745154
1132               Cinema Paradiso  1988  ...      14.177  7.744878
19901                     Paperman  2012  ...     7.19863  7.713951
37863                  Sing Street  2016  ...   10.672862  7.689483
882                  The Apartment  1960  ...     11.9943  7.599317
38718               The Handmaiden  2016  ...   16.727405  7.566166

[10 rows x 6 columns]

In [ ]:
build_chart('Action').head(10)

title  ...        wr
15480                                          Inception  ...  7.955099
12481                                    The Dark Knight  ...  7.948610
4863   The Lord of the Rings: The Fellowship of the Ring  ...  7.929579
7000       The Lord of the Rings: The Return of the King  ...  7.924031
5814               The Lord of the Rings: The Two Towers  ...  7.918382
256                                            Star Wars  ...  7.908327
1154                             The Empire Strikes Back  ...  7.896841
4135                                            Scarface  ...  7.802046
9430                                              Oldboy  ...  7.711649
1910                                       Seven Samurai  ...  7.426145

[10 rows x 6 columns]

### Explain

In [ ]:
genre='Romance'
percentile=0.85

df = gen_md[gen_md['genre'] == genre]
print(gen_md.shape)
print(df.shape)
df.head(2)

(93548, 25)
(6735, 25)


adult                              belongs_to_collection  ...  year    genre
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...  1995  Romance
3  False                                                NaN  ...  1995  Romance

[2 rows x 25 columns]

In [ ]:
# convert vote_count to int
vote_counts = df[df['vote_count'].notnull()]['vote_count'].astype('int')
vote_counts

2         92
3         34
6        141
10       199
16       364
        ... 
45423      8
45425      4
45428      7
45437    146
45460     26
Name: vote_count, Length: 6735, dtype: int64

In [ ]:
# convert vote_averages to int
vote_averages = df[df['vote_average'].notnull()]['vote_average'].astype('int')
vote_averages

2        6
3        6
6        6
10       6
16       7
        ..
45423    5
45425    6
45428    5
45437    8
45460    5
Name: vote_average, Length: 6735, dtype: int64

In [ ]:
C = vote_averages.mean()
m = vote_counts.quantile(percentile)
print(C) # trung bình review rating la 5.4
print(m) # khoảng 85% số lượng vote từ 90 trở lên, tối thiểu số này là OK

5.4041573867854495
90.89999999999964


In [ ]:
qualified = df[(df['vote_count'] >= m) & (df['vote_count'].notnull()) & (df['vote_average'].notnull())][['title', 'year', 'vote_count', 'vote_average', 'popularity']]
print(df.shape) # romance movie only
print(qualified.shape) # roman movie with vote > 90 , average and vote not null

(6735, 25)
(1011, 5)


In [ ]:
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified['wr'] = qualified.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) + (m/(m+x['vote_count']) * C), axis=1)
qualified.head(2)

title  year  vote_count  vote_average popularity        wr
2  Grumpier Old Men  1995          92             6    11.7129  5.703870
6           Sabrina  1995         141             6    6.67728  5.766442

In [ ]:
qualified = qualified.sort_values('wr', ascending=False).head(250)
qualified.head()

title  year  ...  popularity        wr
10309  Dilwale Dulhania Le Jayenge  1995  ...      34.457  8.565285
351                   Forrest Gump  1994  ...     48.3072  7.971357
876                        Vertigo  1958  ...     18.2082  7.811667
40251                   Your Name.  2016  ...   34.461252  7.789489
883               Some Like It Hot  1959  ...     11.8451  7.745154

[5 rows x 6 columns]

### End

## Method 3: suggestion by genre, duration, year

In [ ]:
df=movies_metadata.copy()
#Only keep those features that we require 
df = df[['title','genres', 'release_date', 'runtime', 'vote_average', 'vote_count']]

df.head(2)

title  ... vote_count
0  Toy Story  ...     5415.0
1    Jumanji  ...     2413.0

[2 rows x 6 columns]

In [ ]:
#Convert release_date into pandas datetime format
df['release_date'] = pd.to_datetime(df['release_date'], errors='coerce')

#Extract year from the datetime
df['year'] = df['release_date'].apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)
df.head()

title  ...  year
0                    Toy Story  ...  1995
1                      Jumanji  ...  1995
2             Grumpier Old Men  ...  1995
3            Waiting to Exhale  ...  1995
4  Father of the Bride Part II  ...  1995

[5 rows x 7 columns]

In [ ]:
#Helper function to convert NaT to 0 and all other years to integers.
def convert_int(x):
    try:
        return int(x)
    except:
        return 0

In [ ]:
#Apply convert_int to the year feature
df['year'] = df['year'].apply(convert_int)

In [ ]:
#Drop the release_date column
df = df.drop('release_date', axis=1)

In [ ]:
#Print genres of the first movie
df.iloc[0]['genres']

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

In [ ]:
#Import the literal_eval function from ast
from ast import literal_eval
#Convert all NaN into stringified empty lists
df['genres'] = df['genres'].fillna('[]')

#Apply literal_eval to convert stringified empty lists to the list object
df['genres'] = df['genres'].apply(literal_eval)

#Convert list of dictionaries to a list of strings
df['genres'] = df['genres'].apply(lambda x: [i['name'].lower() for i in x] if isinstance(x, list) else [])

In [ ]:
df

title  ...  year
0                        Toy Story  ...  1995
1                          Jumanji  ...  1995
2                 Grumpier Old Men  ...  1995
3                Waiting to Exhale  ...  1995
4      Father of the Bride Part II  ...  1995
...                            ...  ...   ...
45461                       Subdue  ...     0
45462          Century of Birthing  ...  2011
45463                     Betrayal  ...  2003
45464             Satan Triumphant  ...  1917
45465                     Queerama  ...  2017

[45466 rows x 6 columns]

In [ ]:
#Create a new feature by exploding genres
s = df.apply(lambda x: pd.Series(x['genres']),axis=1).stack().reset_index(level=1, drop=True)

#Name the new feature as 'genre'
s.name = 'genre'

#Create a new dataframe gen_df which by dropping the old 'genres' feature and adding the new 'genre'.
gen_df = df.drop('genres', axis=1).join(s)

In [ ]:
#Print the head of the new gen_df
gen_df

title  runtime  vote_average  vote_count  year      genre
0             Toy Story     81.0           7.7      5415.0  1995  animation
0             Toy Story     81.0           7.7      5415.0  1995     comedy
0             Toy Story     81.0           7.7      5415.0  1995     family
1               Jumanji    104.0           6.9      2413.0  1995  adventure
1               Jumanji    104.0           6.9      2413.0  1995    fantasy
...                 ...      ...           ...         ...   ...        ...
45463          Betrayal     90.0           3.8         6.0  2003     action
45463          Betrayal     90.0           3.8         6.0  2003      drama
45463          Betrayal     90.0           3.8         6.0  2003   thriller
45464  Satan Triumphant     87.0           0.0         0.0  1917        NaN
45465          Queerama     75.0           0.0         0.0  2017        NaN

[93548 rows x 6 columns]

In [ ]:
def build_chart(gen_df, percentile=0.8):
    #Ask for preferred genres
    print("Input preferred genre")
    genre = input()
    
    #Ask for lower limit of duration
    print("Input shortest duration")
    low_time = int(input())
    
    #Ask for upper limit of duration
    print("Input longest duration")
    high_time = int(input())
    
    #Ask for lower limit of timeline
    print("Input earliest year")
    low_year = int(input())
    
    #Ask for upper limit of timeline
    print("Input latest year")
    high_year = int(input())
    
    #Define a new movies variable to store the preferred movies. Copy the contents of gen_df to movies
    movies = gen_df.copy()
    
    #Filter based on the condition
    movies = movies[(movies['genre'] == genre) & 
                    (movies['runtime'] >= low_time) & 
                    (movies['runtime'] <= high_time) & 
                    (movies['year'] >= low_year) & 
                    (movies['year'] <= high_year)]
    
    #Compute the values of C and m for the filtered movies
    C = movies['vote_average'].mean()
    m = movies['vote_count'].quantile(percentile)
    
    #Only consider movies that have higher than m votes. Save this in a new dataframe q_movies
    q_movies = movies.copy().loc[movies['vote_count'] >= m]
    
    #Calculate score using the IMDB formula
    q_movies['score'] = q_movies.apply(lambda x: (x['vote_count']/(x['vote_count']+m) * x['vote_average']) 
                                       + (m/(m+x['vote_count']) * C)
                                       ,axis=1)

    #Sort movies in descending order of their scores
    q_movies = q_movies.sort_values('score', ascending=False)
    
    return q_movies

In [ ]:
gen_df.describe()

runtime  vote_average    vote_count          year
count  93194.000000  93527.000000  93527.000000  93548.000000
mean      96.753375      5.703193    146.629262   1989.223885
std       36.098436      1.743848    592.051356     77.263941
min        0.000000      0.000000      0.000000      0.000000
25%       87.000000      5.100000      4.000000   1979.000000
50%       96.000000      6.000000     13.000000   2001.000000
75%      108.000000      6.800000     48.000000   2010.000000
max     1256.000000     10.000000  14075.000000   2020.000000

In [ ]:
gen_df.genre.value_counts()

drama                                    20265
comedy                                   13182
thriller                                  7624
romance                                   6735
action                                    6596
horror                                    4673
crime                                     4307
documentary                               3932
adventure                                 3496
science fiction                           3049
family                                    2770
mystery                                   2467
fantasy                                   2313
animation                                 1935
foreign                                   1622
music                                     1598
history                                   1398
war                                       1323
western                                   1042
tv movie                                   767
the cartel                                   1
sentai filmwo

In [ ]:
#Generate the chart for top animation movies and display top 5.
build_chart(gen_df).head()

Input preferred genre
comedy
Input shortest duration
60
Input longest duration
120
Input earliest year
2005
Input latest year
2010


title  runtime  vote_average  ...  year   genre     score
13724                    Up     96.0           7.8  ...  2009  comedy  7.762541
15348           Toy Story 3    103.0           7.6  ...  2010  comedy  7.549669
11567           Ratatouille    111.0           7.5  ...  2007  comedy  7.450228
14310          Mary and Max     92.0           7.8  ...  2009  comedy  7.428075
11073  Little Miss Sunshine    102.0           7.5  ...  2006  comedy  7.375567

[5 rows x 7 columns]

In [ ]:
#Generate the chart for top animation movies and display top 5.
build_chart(gen_df).head()

Input preferred genre
adventure
Input shortest duration
60
Input longest duration
120
Input earliest year
2015
Input latest year
2020


title  runtime  vote_average  ...  year      genre     score
36253            Zootopia    108.0           7.7  ...  2016  adventure  7.435592
26564            Deadpool    108.0           7.4  ...  2016  adventure  7.294943
38798   Captain Fantastic    119.0           7.9  ...  2016  adventure  7.187180
26553  Mad Max: Fury Road    120.0           7.3  ...  2015  adventure  7.184099
41457               Moana    107.0           7.3  ...  2016  adventure  7.016132

[5 rows x 7 columns]

# Content Based Recommender

The recommender we built in the previous section suffers some severe limitations. For one, it gives the same recommendation to everyone, regardless of the user's personal taste. If a person who loves romantic movies (and hates action) were to look at our Top 15 Chart, s/he wouldn't probably like most of the movies. If s/he were to go one step further and look at our charts by genre, s/he wouldn't still be getting the best recommendations.

For instance, consider a person who loves *Dilwale Dulhania Le Jayenge*, *My Name is Khan* and *Kabhi Khushi Kabhi Gham*. One inference we can obtain is that the person loves the actor Shahrukh Khan and the director Karan Johar. Even if s/he were to access the romance chart, s/he wouldn't find these as the top recommendations.

To personalise our recommendations more, I am going to build an engine that computes similarity between movies based on certain metrics and suggests movies that are most similar to a particular movie that a user liked. Since we will be using movie metadata (or content) to build this engine, this also known as **Content Based Filtering.**

I will build two Content Based Recommenders based on:
* Movie Overviews and Taglines
* Movie Cast, Crew, Keywords and Genre

Also, as mentioned in the introduction, I will be using a subset of all the movies available to us due to limiting computing power available to me. 

In [ ]:
links_small = pd.read_csv('/content/drive/My Drive/Data/Movie dataset/links_small.csv')
links_small

movieId   imdbId    tmdbId
0           1   114709     862.0
1           2   113497    8844.0
2           3   113228   15602.0
3           4   114885   31357.0
4           5   113041   11862.0
...       ...      ...       ...
9120   162672  3859980  402672.0
9121   163056  4262980  315011.0
9122   163949  2531318  391698.0
9123   164977    27660  137608.0
9124   164979  3447228  410803.0

[9125 rows x 3 columns]

In [ ]:
links_small = links_small[links_small['tmdbId'].notnull()]['tmdbId'].astype('int')
links_small

0          862
1         8844
2        15602
3        31357
4        11862
         ...  
9120    402672
9121    315011
9122    391698
9123    137608
9124    410803
Name: tmdbId, Length: 9112, dtype: int64

In [ ]:
md.id

0           862
1          8844
2         15602
3         31357
4         11862
          ...  
45461    439050
45462    111109
45463     67758
45464    227506
45465    461257
Name: id, Length: 45466, dtype: object

In [ ]:
not_convert_list=[]
for i in range(len(md)):
  try:
    md['id'][i]=int(md['id'][i])
  except:
    print('cannot convert: ',md['id'][i])
    not_convert_list.append(md.index[i])

cannot convert:  1997-08-20
cannot convert:  2012-09-29
cannot convert:  2014-01-01


In [ ]:
not_convert_list

[19730, 29503, 35587]

In [ ]:
md = md.drop([19730, 29503, 35587])

In [ ]:
#Check EDA Notebook for how and why I got these indices.
md['id'] = md['id'].astype('int')

In [ ]:
print(md.shape)
smd = md[md['id'].isin(links_small)]
smd.shape

(45463, 25)


(9099, 25)

We have **9099** movies avaiable in our small movies metadata dataset which is 5 times smaller than our original dataset of 45000 movies.

## Movie Description Based Recommender

Let us first try to build a recommender using movie descriptions and taglines. We do not have a quantitative metric to judge our machine's performance so this will have to be done qualitatively.

In [ ]:
smd.head(5)

index  adult  ...  year                                        description
0      0  False  ...  1995  Led by Woody, Andy's toys live happily in his ...
1      1  False  ...  1995  When siblings Judy and Peter discover an encha...
2      2  False  ...  1995  A family wedding reignites the ancient feud be...
3      3  False  ...  1995  Cheated on, mistreated and stepped on, the wom...
4      4  False  ...  1995  Just when George Banks has recovered from his ...

[5 rows x 27 columns]

In [ ]:
smd['tagline'] = smd['tagline'].fillna('')
smd['description'] = smd['overview'] + smd['tagline']
smd['description'] = smd['description'].fillna('')

In [ ]:
smd['description'] # tag line + overview

0        Led by Woody, Andy's toys live happily in his ...
1        When siblings Judy and Peter discover an encha...
2        A family wedding reignites the ancient feud be...
3        Cheated on, mistreated and stepped on, the wom...
4        Just when George Banks has recovered from his ...
                               ...                        
40224    From the mind behind Evangelion comes a hit la...
40503    The band stormed Europe in 1963, and, in 1964,...
44821    When Molly Hale's sadness of her father's disa...
44826    All your favorite Pokémon characters are back...
45265    While holidaying in the French Alps, a Swedish...
Name: description, Length: 9099, dtype: object

In [ ]:
tf = TfidfVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
tfidf_matrix = tf.fit_transform(smd['description'])

In [ ]:
tfidf_matrix.shape

(9099, 268961)

#### Cosine Similarity

I will be using the Cosine Similarity to calculate a numeric quantity that denotes the similarity between two movies. Mathematically, it is defined as follows:

$cosine(x,y) = \frac{x. y^\intercal}{||x||.||y||} $

Since we have used the TF-IDF Vectorizer, calculating the Dot Product will directly give us the Cosine Similarity Score. Therefore, we will use sklearn's **linear_kernel** instead of cosine_similarities since it is much faster.

In [ ]:
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [ ]:
cosine_sim.shape

(9099, 9099)

In [ ]:
cosine_sim

array([[1.        , 0.00680459, 0.        , ..., 0.        , 0.00289435,
        0.        ],
       [0.00680459, 1.        , 0.01531013, ..., 0.00354933, 0.00640546,
        0.        ],
       [0.        , 0.01531013, 1.        , ..., 0.        , 0.00240445,
        0.00473472],
       ...,
       [0.        , 0.00354933, 0.        , ..., 1.        , 0.14629167,
        0.        ],
       [0.00289435, 0.00640546, 0.00240445, ..., 0.14629167, 1.        ,
        0.        ],
       [0.        , 0.        , 0.00473472, ..., 0.        , 0.        ,
        1.        ]])

We now have a pairwise cosine similarity matrix for all the movies in our dataset. The next step is to write a function that returns the 30 most similar movies based on the cosine similarity score.

In [ ]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])
indices

title
Toy Story                                                0
Jumanji                                                  1
Grumpier Old Men                                         2
Waiting to Exhale                                        3
Father of the Bride Part II                              4
                                                      ... 
Shin Godzilla                                         9094
The Beatles: Eight Days a Week - The Touring Years    9095
Pokémon: Spell of the Unknown                        9096
Pokémon 4Ever: Celebi - Voice of the Forest          9097
Force Majeure                                         9098
Length: 9099, dtype: int64

In [ ]:
def get_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:31]
    movie_indices = [i[0] for i in sim_scores]
    return titles.iloc[movie_indices]

We're all set. Let us now try and get the top recommendations for a few movies and see how good the recommendations are.

In [ ]:
get_recommendations('The Godfather').head(10)

973      The Godfather: Part II
8387                 The Family
3509                       Made
4196         Johnny Dangerously
29               Shanghai Triad
5667                       Fury
2412             American Movie
1582    The Godfather: Part III
4221                    8 Women
2159              Summer of Sam
Name: title, dtype: object

In [ ]:
get_recommendations('The Dark Knight').head(10)

7931                      The Dark Knight Rises
132                              Batman Forever
1113                             Batman Returns
8227    Batman: The Dark Knight Returns, Part 2
7565                 Batman: Under the Red Hood
524                                      Batman
7901                           Batman: Year One
2579               Batman: Mask of the Phantasm
8165    Batman: The Dark Knight Returns, Part 1
2696                                        JFK
Name: title, dtype: object

Explain

In [ ]:
indices

title
Toy Story                                                0
Jumanji                                                  1
Grumpier Old Men                                         2
Waiting to Exhale                                        3
Father of the Bride Part II                              4
                                                      ... 
Shin Godzilla                                         9094
The Beatles: Eight Days a Week - The Touring Years    9095
Pokémon: Spell of the Unknown                         9096
Pokémon 4Ever: Celebi - Voice of the Forest           9097
Force Majeure                                         9098
Length: 9099, dtype: int64

In [ ]:
cosine_sim

array([[1.        , 0.00680476, 0.        , ..., 0.        , 0.00344913,
        0.        ],
       [0.00680476, 1.        , 0.01531062, ..., 0.00357057, 0.00762326,
        0.        ],
       [0.        , 0.01531062, 1.        , ..., 0.        , 0.00286535,
        0.00472155],
       ...,
       [0.        , 0.00357057, 0.        , ..., 1.        , 0.07811616,
        0.        ],
       [0.00344913, 0.00762326, 0.00286535, ..., 0.07811616, 1.        ,
        0.        ],
       [0.        , 0.        , 0.00472155, ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
title='The Godfather'
idx = indices[title]
print('idx: ',idx)
sim_scores = list(enumerate(cosine_sim[idx])) # return (column/film id and value) tại row idx
print('sim_scores: ',sim_scores)
sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True) # sort value
print('sim_scores sorted: ',sim_scores)
sim_scores = sim_scores[1:31]
movie_indices = [i[0] for i in sim_scores]
print('movie_indices: ',movie_indices)
print(titles.iloc[movie_indices])# in tat ca ten phim co list id như trên

idx:  692
sim_scores:  [(0, 0.0), (1, 0.00282712634654008), (2, 0.013274406850346377), (3, 0.0), (4, 0.001911923808690747), (5, 0.009831202255696751), (6, 0.0), (7, 0.0048937541840076175), (8, 0.0), (9, 0.007200486377613863), (10, 0.0), (11, 0.0), (12, 0.002629662427098924), (13, 0.01655348503181121), (14, 0.0), (15, 0.003311321356213731), (16, 0.0), (17, 0.0017852909522175192), (18, 0.0), (19, 0.0), (20, 0.00138960229604634), (21, 0.02347433335488712), (22, 0.002909885967091068), (23, 0.0), (24, 0.0), (25, 0.0), (26, 0.0), (27, 0.011643868586706513), (28, 0.0), (29, 0.056141832299658065), (30, 0.0), (31, 0.0028878395660696552), (32, 0.0), (33, 0.0), (34, 0.0), (35, 0.0038924168745627535), (36, 0.0), (37, 0.0), (38, 0.02717387932966191), (39, 0.0), (40, 0.0), (41, 0.0020748439693151896), (42, 0.0), (43, 0.0), (44, 0.002612852388514977), (45, 0.0), (46, 0.01116205834731158), (47, 0.0), (48, 0.017220777855400132), (49, 0.003365386412223826), (50, 0.006507663461137723), (51, 0.00663189538

End

We see that for **The Dark Knight**, our system is able to identify it as a Batman film and subsequently recommend other Batman films as its top recommendations. But unfortunately, that is all this system can do at the moment. This is not of much use to most people as it doesn't take into considerations very important features such as cast, crew, director and genre, which determine the rating and the popularity of a movie. Someone who liked **The Dark Knight** probably likes it more because of Nolan and would hate **Batman Forever** and every other substandard movie in the Batman Franchise.

Therefore, we are going to use much more suggestive metadata than **Overview** and **Tagline**. In the next subsection, we will build a more sophisticated recommender that takes **genre**, **keywords**, **cast** and **crew** into consideration.

## Metadata Based Recommender

To build our standard metadata based content recommender, we will need to merge our current dataset with the crew and the keyword datasets. Let us prepare this data as our first step.

In [ ]:
credits = pd.read_csv('/content/drive/My Drive/Data/Movie dataset/credits.csv')
keywords = pd.read_csv('/content/drive/My Drive/Data/Movie dataset/keywords.csv')

In [ ]:
credits

cast  ...      id
0      [{'cast_id': 14, 'character': 'Woody (voice)',...  ...     862
1      [{'cast_id': 1, 'character': 'Alan Parrish', '...  ...    8844
2      [{'cast_id': 2, 'character': 'Max Goldman', 'c...  ...   15602
3      [{'cast_id': 1, 'character': "Savannah 'Vannah...  ...   31357
4      [{'cast_id': 1, 'character': 'George Banks', '...  ...   11862
...                                                  ...  ...     ...
45471  [{'cast_id': 0, 'character': '', 'credit_id': ...  ...  439050
45472  [{'cast_id': 1002, 'character': 'Sister Angela...  ...  111109
45473  [{'cast_id': 6, 'character': 'Emily Shaw', 'cr...  ...   67758
45474  [{'cast_id': 2, 'character': '', 'credit_id': ...  ...  227506
45475                                                 []  ...  461257

[45476 rows x 3 columns]

In [ ]:
print(credits.cast[0]) # cac dien vien trong phim (id, nhan vat, gioi tinh, ten)
print(credits.crew[0]) # cac thanh vien trong doan lam phim (dao dien, nha san xuat...)

[{'cast_id': 14, 'character': 'Woody (voice)', 'credit_id': '52fe4284c3a36847f8024f95', 'gender': 2, 'id': 31, 'name': 'Tom Hanks', 'order': 0, 'profile_path': '/pQFoyx7rp09CJTAb932F2g8Nlho.jpg'}, {'cast_id': 15, 'character': 'Buzz Lightyear (voice)', 'credit_id': '52fe4284c3a36847f8024f99', 'gender': 2, 'id': 12898, 'name': 'Tim Allen', 'order': 1, 'profile_path': '/uX2xVf6pMmPepxnvFWyBtjexzgY.jpg'}, {'cast_id': 16, 'character': 'Mr. Potato Head (voice)', 'credit_id': '52fe4284c3a36847f8024f9d', 'gender': 2, 'id': 7167, 'name': 'Don Rickles', 'order': 2, 'profile_path': '/h5BcaDMPRVLHLDzbQavec4xfSdt.jpg'}, {'cast_id': 17, 'character': 'Slinky Dog (voice)', 'credit_id': '52fe4284c3a36847f8024fa1', 'gender': 2, 'id': 12899, 'name': 'Jim Varney', 'order': 3, 'profile_path': '/eIo2jVVXYgjDtaHoF19Ll9vtW7h.jpg'}, {'cast_id': 18, 'character': 'Rex (voice)', 'credit_id': '52fe4284c3a36847f8024fa5', 'gender': 2, 'id': 12900, 'name': 'Wallace Shawn', 'order': 4, 'profile_path': '/oGE6JqPP2xH4tN

In [ ]:
keywords

id                                           keywords
0         862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1        8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2       15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3       31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4       11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...
...       ...                                                ...
46414  439050             [{'id': 10703, 'name': 'tragic love'}]
46415  111109  [{'id': 2679, 'name': 'artist'}, {'id': 14531,...
46416   67758                                                 []
46417  227506                                                 []
46418  461257                                                 []

[46419 rows x 2 columns]

In [ ]:
credits[credits.id==862] # Toy story film

cast  ...   id
0  [{'cast_id': 14, 'character': 'Woody (voice)',...  ...  862

[1 rows x 3 columns]

In [ ]:
keywords.keywords[0] # Toy story film

"[{'id': 931, 'name': 'jealousy'}, {'id': 4290, 'name': 'toy'}, {'id': 5202, 'name': 'boy'}, {'id': 6054, 'name': 'friendship'}, {'id': 9713, 'name': 'friends'}, {'id': 9823, 'name': 'rivalry'}, {'id': 165503, 'name': 'boy next door'}, {'id': 170722, 'name': 'new toy'}, {'id': 187065, 'name': 'toy comes to life'}]"

In [ ]:
keywords['id'] = keywords['id'].astype('int')
credits['id'] = credits['id'].astype('int')
md['id'] = md['id'].astype('int')

In [ ]:
md.head(3)

adult                              belongs_to_collection  ... vote_count  year
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  ...     5415.0  1995
1  False                                                NaN  ...     2413.0  1995
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...       92.0  1995

[3 rows x 25 columns]

In [ ]:
md.shape

(45463, 25)

In [ ]:
md = md.merge(credits, on='id')
md = md.merge(keywords, on='id')

In [ ]:
md.head(3)

adult  ...                                           keywords
0  False  ...  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1  False  ...  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  False  ...  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...

[3 rows x 28 columns]

In [ ]:
md.shape

(46628, 28)

In [ ]:
smd = md[md['id'].isin(links_small)]
smd.shape # loc 45000 phim thanh 9219 phim thôi

(9219, 28)

We now have our cast, crew, genres and credits, all in one dataframe. Let us wrangle this a little more using the following intuitions:

1. **Crew:** From the crew, we will only pick the director as our feature since the others don't contribute that much to the *feel* of the movie.
2. **Cast:** Choosing Cast is a little more tricky. Lesser known actors and minor roles do not really affect people's opinion of a movie. Therefore, we must only select the major characters and their respective actors. Arbitrarily we will choose the top 3 actors that appear in the credits list. 

In [ ]:
smd.head(2)

adult  ...                                           keywords
0  False  ...  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1  False  ...  [{'id': 10090, 'name': 'board game'}, {'id': 1...

[2 rows x 28 columns]

from ast import literal_eval

In [ ]:
smd['cast'] = smd['cast'].apply(literal_eval)
smd['crew'] = smd['crew'].apply(literal_eval)
smd['keywords'] = smd['keywords'].apply(literal_eval)
smd['cast_size'] = smd['cast'].apply(lambda x: len(x))
smd['crew_size'] = smd['crew'].apply(lambda x: len(x))

In [ ]:
smd.head(2)

adult  ... crew_size
0  False  ...       106
1  False  ...        16

[2 rows x 30 columns]

In [ ]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [ ]:
smd['director'] = smd['crew'].apply(get_director)

In [ ]:
smd['cast'] = smd['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])
smd['cast'] = smd['cast'].apply(lambda x: x[:3] if len(x) >=3 else x)

In [ ]:
smd['keywords'] = smd['keywords'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

In [ ]:
smd.head(2)

adult  ...       director
0  False  ...  John Lasseter
1  False  ...   Joe Johnston

[2 rows x 31 columns]

My approach to building the recommender is going to be extremely *hacky*. What I plan on doing is creating a metadata dump for every movie which consists of **genres, director, main actors and keywords.** I then use a **Count Vectorizer** to create our count matrix as we did in the Description Recommender. The remaining steps are similar to what we did earlier: we calculate the cosine similarities and return movies that are most similar.

These are steps I follow in the preparation of my genres and credits data:
1. **Strip Spaces and Convert to Lowercase** from all our features. This way, our engine will not confuse between **Johnny Depp** and **Johnny Galecki.** 
2. **Mention Director 3 times** to give it more weight relative to the entire cast.

In [ ]:
smd['cast'] = smd['cast'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [ ]:
smd['director'] = smd['director'].astype('str').apply(lambda x: str.lower(x.replace(" ", "")))
smd['director'] = smd['director'].apply(lambda x: [x,x, x])

In [ ]:
smd.head(2)

adult  ...                                    director
0  False  ...  [johnlasseter, johnlasseter, johnlasseter]
1  False  ...     [joejohnston, joejohnston, joejohnston]

[2 rows x 31 columns]

#### Keywords

We will do a small amount of pre-processing of our keywords before putting them to any use. As a first step, we calculate the frequenct counts of every keyword that appears in the dataset.

In [ ]:
s = smd.apply(lambda x: pd.Series(x['keywords']),axis=1).stack().reset_index(level=1, drop=True)
s.name = 'keyword'

In [ ]:
s

0           jealousy
0                toy
0                boy
0         friendship
0            friends
            ...     
41391    destruction
41391          kaiju
41391          toyko
41669          music
41669    documentary
Name: keyword, Length: 64407, dtype: object

In [ ]:
s = s.value_counts()
s[:5]

independent film        610
woman director          550
murder                  399
duringcreditsstinger    327
based on novel          318
Name: keyword, dtype: int64

In [ ]:
s.shape

(12940,)

Keywords occur in frequencies ranging from 1 to 610. We do not have any use for keywords that occur only once. Therefore, these can be safely removed. Finally, we will convert every word to its stem so that words such as *Dogs* and *Dog* are considered the same.

In [ ]:
s = s[s > 1]
s

independent film         610
woman director           550
murder                   399
duringcreditsstinger     327
based on novel           318
                        ... 
strategic air command      2
translation                2
computer expert            2
live concert               2
sexual murder              2
Name: keyword, Length: 6709, dtype: int64

In [ ]:
#from nltk.stem.snowball import SnowballStemmer
stemmer = SnowballStemmer('english')
stemmer.stem('dogs')

'dog'

In [ ]:
def filter_keywords(x):
    words = []
    for i in x:
        if i in s:
            words.append(i)
    return words

In [ ]:
smd['keywords'] = smd['keywords'].apply(filter_keywords)
smd['keywords'] = smd['keywords'].apply(lambda x: [stemmer.stem(i) for i in x])
smd['keywords'] = smd['keywords'].apply(lambda x: [str.lower(i.replace(" ", "")) for i in x])

In [ ]:
smd['soup'] = smd['keywords'] + smd['cast'] + smd['director'] + smd['genres']
smd['soup'] = smd['soup'].apply(lambda x: ' '.join(x))

In [ ]:
smd.head(2)

adult  ...                                               soup
0  False  ...  jealousi toy boy friendship friend rivalri boy...
1  False  ...  boardgam disappear basedonchildren'sbook newho...

[2 rows x 32 columns]

In [ ]:
count = CountVectorizer(analyzer='word',ngram_range=(1, 2),min_df=0, stop_words='english')
count_matrix = count.fit_transform(smd['soup'])

In [ ]:
count_matrix.shape

(9219, 107377)

In [ ]:
cosine_sim = cosine_similarity(count_matrix, count_matrix)

In [ ]:
cosine_sim

array([[1.        , 0.02441931, 0.02738955, ..., 0.        , 0.        ,
        0.        ],
       [0.02441931, 1.        , 0.        , ..., 0.02973505, 0.02500782,
        0.        ],
       [0.02738955, 0.        , 1.        , ..., 0.03335187, 0.        ,
        0.        ],
       ...,
       [0.        , 0.02973505, 0.03335187, ..., 1.        , 0.08700222,
        0.        ],
       [0.        , 0.02500782, 0.        , ..., 0.08700222, 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
cosine_sim.shape

(9219, 9219)

In [ ]:
smd = smd.reset_index()
titles = smd['title']
indices = pd.Series(smd.index, index=smd['title'])

In [ ]:
indices

title
Toy Story                                                0
Jumanji                                                  1
Grumpier Old Men                                         2
Waiting to Exhale                                        3
Father of the Bride Part II                              4
                                                      ... 
The Last Brickmaker in America                        9214
Rustom                                                9215
Mohenjo Daro                                          9216
Shin Godzilla                                         9217
The Beatles: Eight Days a Week - The Touring Years    9218
Length: 9219, dtype: int64

In [ ]:
titles

0                                               Toy Story
1                                                 Jumanji
2                                        Grumpier Old Men
3                                       Waiting to Exhale
4                             Father of the Bride Part II
                              ...                        
9214                       The Last Brickmaker in America
9215                                               Rustom
9216                                         Mohenjo Daro
9217                                        Shin Godzilla
9218    The Beatles: Eight Days a Week - The Touring Y...
Name: title, Length: 9219, dtype: object

We will reuse the get_recommendations function that we had written earlier. Since our cosine similarity scores have changed, we expect it to give us different (and probably better) results. Let us check for **The Dark Knight** again and see what recommendations I get this time around.

In [ ]:
get_recommendations('The Dark Knight').head(10)

8031         The Dark Knight Rises
6218                 Batman Begins
6623                  The Prestige
2085                     Following
7648                     Inception
4145                      Insomnia
3381                       Memento
8613                  Interstellar
7659    Batman: Under the Red Hood
1134                Batman Returns
Name: title, dtype: object

I am much more satisfied with the results I get this time around. The recommendations seem to have recognized other Christopher Nolan movies (due to the high weightage given to director) and put them as top recommendations. I enjoyed watching **The Dark Knight** as well as some of the other ones in the list including **Batman Begins**, **The Prestige** and **The Dark Knight Rises**. 

We can of course experiment on this engine by trying out different weights for our features (directors, actors, genres), limiting the number of keywords that can be used in the soup, weighing genres based on their frequency, only showing movies with the same languages, etc.

Let me also get recommendations for another movie, **Mean Girls** which happens to be my girlfriend's favorite movie.

In [ ]:
get_recommendations('Mean Girls').head(10)

3319               Head Over Heels
4763                 Freaky Friday
1329              The House of Yes
6277              Just Like Heaven
7905         Mr. Popper's Penguins
7332    Ghosts of Girlfriends Past
6959     The Spiderwick Chronicles
8883                      The DUFF
6698         It's a Boy Girl Thing
7377       I Love You, Beth Cooper
Name: title, dtype: object

#### Popularity and Ratings

One thing that we notice about our recommendation system is that it recommends movies regardless of ratings and popularity. It is true that **Batman and Robin** has a lot of similar characters as compared to **The Dark Knight** but it was a terrible movie that shouldn't be recommended to anyone.

Therefore, we will add a mechanism to remove bad movies and return movies which are popular and have had a good critical response.

I will take the top 25 movies based on similarity scores and calculate the vote of the 60th percentile movie. Then, using this as the value of $m$, we will calculate the weighted rating of each movie using IMDB's formula like we did in the Simple Recommender section.

In [ ]:
def improved_recommendations(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:26]
    movie_indices = [i[0] for i in sim_scores]
    
    movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
    vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
    vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
    C = vote_averages.mean()
    m = vote_counts.quantile(0.60)
    qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
    qualified['vote_count'] = qualified['vote_count'].astype('int')
    qualified['vote_average'] = qualified['vote_average'].astype('int')
    qualified['wr'] = qualified.apply(weighted_rating, axis=1)
    qualified = qualified.sort_values('wr', ascending=False).head(10)
    return qualified

In [ ]:
improved_recommendations('The Dark Knight')

title  vote_count  ...  year        wr
7648                           Inception       14075  ...  2010  7.983343
8613                        Interstellar       11187  ...  2014  7.979077
6623                        The Prestige        4510  ...  2006  7.948714
3381                             Memento        4168  ...  2000  7.944596
8031               The Dark Knight Rises        9263  ...  2012  6.984492
6218                       Batman Begins        7511  ...  2005  6.980918
1134                      Batman Returns        1706  ...  1992  5.969858
132                       Batman Forever        1529  ...  1995  5.022679
9024  Batman v Superman: Dawn of Justice        7189  ...  2016  5.005046
1260                      Batman & Robin        1447  ...  1997  4.082995

[10 rows x 5 columns]

### Explain:

In [ ]:
title='The Dark Knight'
idx = indices[title]
sim_scores = list(enumerate(cosine_sim[idx]))
sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
sim_scores = sim_scores[1:26]
movie_indices = [i[0] for i in sim_scores]
movies = smd.iloc[movie_indices][['title', 'vote_count', 'vote_average', 'year']]
movies

title  vote_count  vote_average  year
8031               The Dark Knight Rises      9263.0           7.6  2012
6218                       Batman Begins      7511.0           7.5  2005
6623                        The Prestige      4510.0           8.0  2006
2085                           Following       363.0           7.2  1998
7648                           Inception     14075.0           8.1  2010
4145                            Insomnia      1181.0           6.8  2002
3381                             Memento      4168.0           8.1  2000
8613                        Interstellar     11187.0           8.1  2014
7659          Batman: Under the Red Hood       459.0           7.6  2010
1134                      Batman Returns      1706.0           6.6  1992
8927             Kidnapping Mr. Heineken       193.0           5.8  2015
5943                            Thursday        84.0           7.0  1998
1260                      Batman & Robin      1447.0           4.2  1997
9024  Batman v Superman: Dawn of Justice      7189.0           5.7  2016
4021                The Long Good Friday        87.0           7.1  1980
5809                         Point Blank        97.0           7.1  1967
7362     Gangster's Paradise: Jerusalema        16.0           6.8  2008
7561                         Harry Brown       351.0           6.7  2009
7582                            Defendor       197.0           6.5  2009
8001                    Batman: Year One       255.0           7.1  2011
2754                        Death Wish 3        68.0           6.1  1985
132                       Batman Forever      1529.0           5.2  1995
2131                            Superman      1042.0           6.9  1978
2448                          Nighthawks        87.0           6.4  1981
5098                        The Enforcer        21.0           7.4  1951

In [ ]:

vote_counts = movies[movies['vote_count'].notnull()]['vote_count'].astype('int')
vote_averages = movies[movies['vote_average'].notnull()]['vote_average'].astype('int')
C = vote_averages.mean()
m = vote_counts.quantile(0.60)
qualified = movies[(movies['vote_count'] >= m) & (movies['vote_count'].notnull()) & (movies['vote_average'].notnull())]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified['wr'] = qualified.apply(weighted_rating, axis=1)
qualified = qualified.sort_values('wr', ascending=False).head(10)
qualified

title  vote_count  ...  year        wr
7648                           Inception       14075  ...  2010  7.872621
8613                        Interstellar       11187  ...  2014  7.843131
6623                        The Prestige        4510  ...  2006  7.662461
3381                             Memento        4168  ...  2000  7.641301
8031               The Dark Knight Rises        9263  ...  2012  6.936548
6218                       Batman Begins        7511  ...  2005  6.923913
1134                      Batman Returns        1706  ...  1992  6.206438
132                       Batman Forever        1529  ...  1995  5.676520
9024  Batman v Superman: Dawn of Justice        7189  ...  2016  5.224783
1260                      Batman & Robin        1447  ...  1997  5.167624

[10 rows x 5 columns]

### End

Let me also get the recommendations for **Mean Girls**, my girlfriend's favorite movie.

In [ ]:
improved_recommendations('Mean Girls')

title  vote_count  ...  year        wr
1547                       The Breakfast Club        2189  ...  1985  6.709602
390                        Dazed and Confused         588  ...  1993  6.254682
8883                                 The DUFF        1372  ...  2015  5.818541
3712                     The Princess Diaries        1063  ...  2001  5.781086
4763                            Freaky Friday         919  ...  2003  5.757786
6277                         Just Like Heaven         595  ...  2005  5.681521
6959                The Spiderwick Chronicles         593  ...  2008  5.680901
7494  American Pie Presents: The Book of Love         454  ...  2009  5.119690
7332               Ghosts of Girlfriends Past         716  ...  2009  5.092422
7905                    Mr. Popper's Penguins         775  ...  2011  5.087912

[10 rows x 5 columns]

Unfortunately, **Batman and Robin** does not disappear from our recommendation list. This is probably due to the fact that it is rated a 4, which is only slightly below average on TMDB. It certainly doesn't deserve a 4 when amazing movies like **The Dark Knight Rises** has only a 7. However, there is nothing much we can do about this. Therefore, we will conclude our Content Based Recommender section here and come back to it when we build a hybrid engine.

# Collaborative Filtering

In [ ]:
movies=pd.read_csv('/content/drive/My Drive/Data/Movie dataset/Small/movies.csv',encoding = "ISO-8859-1")
print(movies.shape)
movies.head()

(2500, 3)


movieId  ...                                       genres
0        1  ...  Adventure|Animation|Children|Comedy|Fantasy
1        2  ...                   Adventure|Children|Fantasy
2        3  ...                               Comedy|Romance
3        4  ...                         Comedy|Drama|Romance
4        5  ...                                       Comedy

[5 rows x 3 columns]

In [ ]:
titles=movies[['movieId','title']]
titles.set_index('movieId')

title
movieId                                    
1                          Toy Story (1995)
2                            Jumanji (1995)
3                   Grumpier Old Men (1995)
4                  Waiting to Exhale (1995)
5        Father of the Bride Part II (1995)
...                                     ...
109487                  Interstellar (2014)
111362    X-Men: Days of Future Past (2014)
111759              Edge of Tomorrow (2014)
112556                     Gone Girl (2014)
112852       Guardians of the Galaxy (2014)

[2500 rows x 1 columns]

In [ ]:
ratings=pd.read_csv('/content/drive/My Drive/Data/Movie dataset/Small/ratings.csv',encoding = "ISO-8859-1")
print(ratings.shape)
ratings.head()

(264505, 4)


userId  movieId  rating   timestamp
0   12882        1     4.0  1147195252
1   12882       32     3.5  1147195307
2   12882       47     5.0  1147195343
3   12882       50     5.0  1147185499
4   12882      110     4.5  1147195239

In [ ]:
print(len(ratings.userId.unique()))
print(len(ratings.movieId.unique()))

862
2500


## Normalize rating

In [ ]:
Mean = ratings.groupby(by="userId",as_index=False)['rating'].mean()
Mean.rename(columns={'rating':'avg_rating'},inplace=True)
Mean

userId  avg_rating
0       316    3.329457
1       320    3.701613
2       359    3.685474
3       370    3.794404
4       910    3.898080
..      ...         ...
857  137609    3.711538
858  137805    2.779210
859  138072    2.432927
860  138176    4.107143
861  138200    3.422249

[862 rows x 2 columns]

In [ ]:
rating_avg = pd.merge(ratings,Mean,on='userId')
rating_avg # rating_x: rating in each movie; rating_y: average rating all movie

userId  movieId  rating   timestamp  avg_rating
0        12882        1     4.0  1147195252    4.061321
1        12882       32     3.5  1147195307    4.061321
2        12882       47     5.0  1147195343    4.061321
3        12882       50     5.0  1147185499    4.061321
4        12882      110     4.5  1147195239    4.061321
...        ...      ...     ...         ...         ...
264500   90097    54286     5.0  1336143258    3.963415
264501   90097    60069     4.5  1336144698    3.963415
264502   90097    68358     5.0  1336144031    3.963415
264503   90097    68954     5.0  1336146596    3.963415
264504   90097    73017     4.0  1336148879    3.963415

[264505 rows x 5 columns]

In [ ]:
rating_avg.columns

Index(['userId', 'movieId', 'rating', 'timestamp', 'avg_rating'], dtype='object')

In [ ]:
rating_avg['adg_rating']=rating_avg['rating']-rating_avg['avg_rating']
rating_avg.head() # phim nao co rating cao hon muc trung binh cua 1 user

userId  movieId  rating   timestamp  avg_rating  adg_rating
0   12882        1     4.0  1147195252    4.061321   -0.061321
1   12882       32     3.5  1147195307    4.061321   -0.561321
2   12882       47     5.0  1147195343    4.061321    0.938679
3   12882       50     5.0  1147185499    4.061321    0.938679
4   12882      110     4.5  1147195239    4.061321    0.438679

## User-based Collaborative Filtering

In [ ]:
user_movie_matrix = rating_avg.pivot_table(
    index='userId', 
    columns='movieId', 
    values='adg_rating',
    aggfunc='mean'
)
user_movie_matrix

movieId    1         2         3       ...    111759    112556    112852
userId                                 ...                              
316     -0.829457       NaN       NaN  ...       NaN       NaN       NaN
320           NaN       NaN       NaN  ...       NaN       NaN       NaN
359      1.314526       NaN       NaN  ...       NaN       NaN       NaN
370      0.705596  0.205596       NaN  ...  0.705596 -0.294404 -0.794404
910      1.101920  0.101920 -0.398080  ...  0.601920       NaN       NaN
...           ...       ...       ...  ...       ...       ...       ...
137609        NaN       NaN       NaN  ...       NaN       NaN       NaN
137805   1.220790 -0.779210       NaN  ...       NaN       NaN  1.220790
138072   0.567073 -1.432927 -1.432927  ...       NaN       NaN       NaN
138176        NaN       NaN       NaN  ...       NaN       NaN       NaN
138200  -0.422249       NaN       NaN  ...       NaN       NaN       NaN

[862 rows x 2500 columns]

In [ ]:
# Replacing NaN by user Average (fill na by row)
user_movie_matrix = user_movie_matrix.apply(lambda row: row.fillna(row.mean()), axis=1)

In [ ]:
user_movie_matrix.head()

movieId        1             2       ...        112556        112852
userId                               ...                            
316     -8.294574e-01  1.893404e-16  ...  1.893404e-16  1.893404e-16
320      4.297638e-17  4.297638e-17  ...  4.297638e-17  4.297638e-17
359      1.314526e+00 -1.135546e-16  ... -1.135546e-16 -1.135546e-16
370      7.055961e-01  2.055961e-01  ... -2.944039e-01 -7.944039e-01
910      1.101920e+00  1.019202e-01  ...  6.795811e-16  6.795811e-16

[5 rows x 2500 columns]

In [ ]:
b=cosine_similarity(user_movie_matrix)
b

array([[ 1.00000000e+00,  6.00630917e-02,  7.20754753e-02, ...,
         7.73765811e-02,  2.69240918e-02, -2.27270553e-02],
       [ 6.00630917e-02,  1.00000000e+00,  6.30541209e-02, ...,
         9.88451426e-02,  3.87524588e-02,  5.66393007e-02],
       [ 7.20754753e-02,  6.30541209e-02,  1.00000000e+00, ...,
         1.02698453e-01,  9.93065563e-02,  3.14714649e-03],
       ...,
       [ 7.73765811e-02,  9.88451426e-02,  1.02698453e-01, ...,
         1.00000000e+00,  9.49195502e-02,  6.47542404e-02],
       [ 2.69240918e-02,  3.87524588e-02,  9.93065563e-02, ...,
         9.49195502e-02,  1.00000000e+00,  8.99631810e-04],
       [-2.27270553e-02,  5.66393007e-02,  3.14714649e-03, ...,
         6.47542404e-02,  8.99631810e-04,  1.00000000e+00]])

In [ ]:
np.fill_diagonal(b, 0 )
# Chuyen toan bo hang cheo 1 thanh 0 (user trung nhau khong co y nghia o day)

In [ ]:
user_user_sim_matrix = pd.DataFrame(b)
user_user_sim_matrix.head()

0         1         2    ...       859       860       861
0  0.000000  0.060063  0.072075  ...  0.077377  0.026924 -0.022727
1  0.060063  0.000000  0.063054  ...  0.098845  0.038752  0.056639
2  0.072075  0.063054  0.000000  ...  0.102698  0.099307  0.003147
3  0.043266  0.027315  0.135836  ...  0.052187  0.050997  0.009950
4  0.039305  0.006811  0.076131  ...  0.073272 -0.012058  0.007610

[5 rows x 862 columns]

In [ ]:
user_user_sim_matrix.columns = user_movie_matrix.index
user_user_sim_matrix['userId'] = user_movie_matrix.index
user_user_sim_matrix = user_user_sim_matrix.set_index('userId')
user_user_sim_matrix.head()

userId    316       320       359     ...    138072    138176    138200
userId                                ...                              
316     0.000000  0.060063  0.072075  ...  0.077377  0.026924 -0.022727
320     0.060063  0.000000  0.063054  ...  0.098845  0.038752  0.056639
359     0.072075  0.063054  0.000000  ...  0.102698  0.099307  0.003147
370     0.043266  0.027315  0.135836  ...  0.052187  0.050997  0.009950
910     0.039305  0.006811  0.076131  ...  0.073272 -0.012058  0.007610

[5 rows x 862 columns]

### Top 30 neighbours for each user

In [ ]:
df=user_user_sim_matrix.copy()
print(df.shape)
df.head()

(862, 862)


userId    316       320       359     ...    138072    138176    138200
userId                                ...                              
316     0.000000  0.060063  0.072075  ...  0.077377  0.026924 -0.022727
320     0.060063  0.000000  0.063054  ...  0.098845  0.038752  0.056639
359     0.072075  0.063054  0.000000  ...  0.102698  0.099307  0.003147
370     0.043266  0.027315  0.135836  ...  0.052187  0.050997  0.009950
910     0.039305  0.006811  0.076131  ...  0.073272 -0.012058  0.007610

[5 rows x 862 columns]

In [ ]:
n=30
order = np.argsort(df.values, axis=1)
print(order.shape)
order
# Returns the indices that would sort an array (index column id only)

(862, 862)


array([[682, 506, 332, ...,  44, 729, 704],
       [694, 720, 332, ..., 178, 704,  64],
       [332, 659, 730, ..., 637, 595, 634],
       ...,
       [691, 793, 332, ..., 207,  52, 417],
       [323, 290, 601, ..., 317, 254, 341],
       [318, 528, 447, ..., 557,  84,  64]])

In [ ]:
df = df.apply(lambda x: pd.Series(x.sort_values(ascending=False).iloc[:n].index, 
                                  index=['top{}'.format(i) for i in range(1, n+1)]),
              axis=1)
# sort value tung row theo thu tu cao xuong thap, return index of each value
print(df.shape)
sim_user_30_u=df
sim_user_30_u.head()
# return column name with highest value (decending)

(862, 30)


top1    top2    top3   top4  ...   top27   top28   top29   top30
userId                                 ...                                
316     113673  117918    9050  12882  ...   58265   89527   49830   63902
320      12288  113673   28159  79846  ...   80946   10055   64365  106512
359     102118   96482  102532  50898  ...   86768  129498  131620   58346
370      46645   42245   40768  23428  ...  117144   96482    2988  108195
910      87042  131620   67352  40768  ...   18115  114601   23428   17039

[5 rows x 30 columns]

### All movie each user

In [ ]:
rating_avg

userId  movieId  rating   timestamp  avg_rating  adg_rating
0        12882        1     4.0  1147195252    4.061321   -0.061321
1        12882       32     3.5  1147195307    4.061321   -0.561321
2        12882       47     5.0  1147195343    4.061321    0.938679
3        12882       50     5.0  1147185499    4.061321    0.938679
4        12882      110     4.5  1147195239    4.061321    0.438679
...        ...      ...     ...         ...         ...         ...
264500   90097    54286     5.0  1336143258    3.963415    1.036585
264501   90097    60069     4.5  1336144698    3.963415    0.536585
264502   90097    68358     5.0  1336144031    3.963415    1.036585
264503   90097    68954     5.0  1336146596    3.963415    1.036585
264504   90097    73017     4.0  1336148879    3.963415    0.036585

[264505 rows x 6 columns]

In [ ]:
rating_avg = rating_avg.astype({"movieId": str})
Movie_user = rating_avg.groupby(by = 'userId')['movieId'].apply(lambda x:','.join(x))
Movie_user


userId
316       1,7,10,32,34,150,160,163,165,185,196,223,253,2...
320       47,50,260,296,508,527,541,648,750,923,1136,119...
359       1,6,10,11,19,22,24,29,32,34,36,42,44,47,50,62,...
370       1,2,6,25,32,47,50,70,111,153,223,293,296,306,3...
910       1,2,3,5,6,11,15,16,25,32,34,52,62,95,100,110,1...
                                ...                        
137609    28,47,252,256,261,318,327,355,562,1042,1246,14...
137805    1,2,10,11,34,39,44,46,47,62,66,70,110,112,141,...
138072    1,2,3,9,10,11,12,18,19,24,31,34,39,44,50,63,65...
138176    44,318,502,527,711,858,1092,1221,1371,2023,239...
138200    1,32,39,47,110,165,215,253,342,355,364,367,374...
Name: movieId, Length: 862, dtype: object

### Make recommendation

In [ ]:
userId=370

#### list movie seen by this user

In [ ]:

Movie_seen_by_user=list(ratings[ratings.userId==userId].movieId.values)
print(len(Movie_seen_by_user))
print(Movie_seen_by_user[:10])

411
[1, 2, 6, 25, 32, 47, 50, 70, 111, 153]


#### Tim tat ca user id gan voi user da chon

In [ ]:
all_user_similar = sim_user_30_u[sim_user_30_u.index==userId].values
all_user_similar # tim tat ca user id gan voi user da chon

array([[ 46645,  42245,  40768,  23428, 123707,  60016,  45120, 113645,
         97195, 102118,  58265, 113540, 102532, 120782,  17039, 117007,
        101137,  57937,  27365,  41244,   5611,  20530,   2702,  38159,
           359,  43354, 117144,  96482,   2988, 108195]])

In [ ]:
all_user_similar = all_user_similar.squeeze().tolist()
print(all_user_similar)

[46645, 42245, 40768, 23428, 123707, 60016, 45120, 113645, 97195, 102118, 58265, 113540, 102532, 120782, 17039, 117007, 101137, 57937, 27365, 41244, 5611, 20530, 2702, 38159, 359, 43354, 117144, 96482, 2988, 108195]


#### Movies_under_consideration

In [ ]:
d = Movie_user[Movie_user.index.isin(all_user_similar)]
d # tim tat ca cac phim cua cac user similar voi user da chon

userId
359       1,6,10,11,19,22,24,29,32,34,36,42,44,47,50,62,...
2702      1,6,12,19,32,44,47,50,110,111,150,153,158,165,...
2988      1,2,18,21,25,29,32,34,36,47,50,69,73,82,85,110...
5611      1,32,47,110,111,145,150,172,193,223,253,260,29...
17039     1,29,32,47,50,111,150,260,296,318,431,541,589,...
20530     1,6,11,21,25,32,47,50,105,110,141,153,185,191,...
23428     1,2,47,50,70,107,150,153,223,231,260,296,318,3...
27365     6,18,32,47,50,70,111,153,223,288,296,316,318,3...
38159     1,2,5,6,10,11,16,19,22,39,42,47,50,61,65,70,10...
40768     1,2,10,13,17,34,44,47,48,50,110,111,150,158,18...
41244     1,7,11,32,34,47,50,60,62,73,104,110,150,153,15...
42245     10,16,21,32,34,42,44,47,50,110,111,150,151,153...
43354     2,6,10,16,17,31,34,47,50,88,160,161,165,170,18...
45120     32,50,76,175,193,215,231,260,293,296,318,319,3...
46645     1,2,10,16,19,32,34,36,47,50,52,58,70,97,110,11...
57937     1,6,10,16,21,32,34,47,50,104,110,111,150,153,1...
58265     6,10,11,15,19,21,32,47,

In [ ]:
l = ','.join(d.values)
Movie_seen_by_similar_users = l.split(',')
print(len(Movie_seen_by_similar_users))
print(len(set(Movie_seen_by_similar_users)))
# List tat ca phim ma nhung similar users đã xem

15313
2268


In [ ]:
Movies_under_consideration = list(set(Movie_seen_by_similar_users)-set(list(map(str, Movie_seen_by_user))))
print(len(Movies_under_consideration))
print(Movies_under_consideration)
# Lấy list movie by similar user - movie seen by user được list cac moive tiềm năng

1857
['3638', '1208', '4867', '2375', '2416', '73', '6942', '350', '432', '3363', '2421', '4223', '2335', '3623', '2973', '3146', '3871', '3681', '4754', '2001', '2072', '4159', '2398', '40278', '940', '261', '207', '3968', '1210', '586', '5064', '7156', '6155', '2163', '1405', '2318', '546', '4039', '7', '902', '6863', '1120', '4047', '109487', '2382', '2360', '2094', '81', '1334', '2606', '3384', '71057', '4214', '2598', '1425', '233', '1556', '4865', '5970', '2707', '38038', '4010', '932', '80489', '5995', '2348', '5294', '2961', '5446', '8977', '47629', '2405', '3639', '2278', '468', '1307', '4386', '5254', '910', '1991', '64622', '585', '5283', '8810', '3896', '3298', '533', '1488', '17', '1848', '2087', '44199', '8370', '524', '1086', '5464', '1188', '2021', '8969', '780', '407', '2408', '7569', '520', '3258', '4247', '502', '2067', '3633', '429', '3916', '46', '3802', '46970', '1213', '8947', '33166', '3246', '82459', '1608', '1894', '4887', '1873', '3809', '1799', '3897', '2947

In [ ]:
Movies_under_consideration = list(map(int, Movies_under_consideration))
print(len(Movies_under_consideration))
# Chuyển list này thành int

1857


#### Tính score từng moive under này (Very hard way)

In [ ]:
Mean.head()

userId  avg_rating
0     316    3.329457
1     320    3.701613
2     359    3.685474
3     370    3.794404
4     910    3.898080

In [ ]:
user_movie_matrix.head()

movieId        1             2       ...        112556        112852
userId                               ...                            
316     -8.294574e-01  1.893404e-16  ...  1.893404e-16  1.893404e-16
320      4.297638e-17  4.297638e-17  ...  4.297638e-17  4.297638e-17
359      1.314526e+00 -1.135546e-16  ... -1.135546e-16 -1.135546e-16
370      7.055961e-01  2.055961e-01  ... -2.944039e-01 -7.944039e-01
910      1.101920e+00  1.019202e-01  ...  6.795811e-16  6.795811e-16

[5 rows x 2500 columns]

In [ ]:
user_user_sim_matrix.head()

userId    316       320       359     ...    138072    138176    138200
userId                                ...                              
316     0.000000  0.060063  0.072075  ...  0.077377  0.026924 -0.022727
320     0.060063  0.000000  0.063054  ...  0.098845  0.038752  0.056639
359     0.072075  0.063054  0.000000  ...  0.102698  0.099307  0.003147
370     0.043266  0.027315  0.135836  ...  0.052187  0.050997  0.009950
910     0.039305  0.006811  0.076131  ...  0.073272 -0.012058  0.007610

[5 rows x 862 columns]

In [ ]:
score = []
for item in Movies_under_consideration:
  #print('-'*50)
  #print('item: ',item)
  c = user_movie_matrix.loc[:,item] 
  #print('c loc ra rating của movie đó theo toàn bộ user \n',c) # loc ra toan bo user theo colum movie dang chọn
  #print(len(c))
  d = c[c.index.isin(b)]  
  #print('d loc cac user id/rating average lien quan voi user da chon\n',d) # loc cac user id lien quan voi user da chon (b)
  #print(len(d))
  f = d[d.notnull()] 
  #print('f loai bo null user\n',f) # loai bo null user
  #print(len(f))
  avg_user = Mean.loc[Mean['userId'] == userId,'avg_rating'].values[0] 
  #print('avg_user\n',avg_user)#loc mean rating value cua user đang chon
  index = f.index.values.squeeze().tolist()
  #print('index return index cac user id similar to user đã chọn\n',index) # return index cac user id similar to user đã chọn
  #print(len(index))
  corr = user_user_sim_matrix.loc[userId,index]
  #print('corr: trong danh sach similarity user x user loc ra row=user, column= list user similarity\n',corr) # trong danh sach similarity user x user loc ra row=user, column= list user similarity
  fin = pd.concat([f, corr], axis=1)
  #print('fin\n',fin)6. 
  fin.columns = ['adg_score','correlation']
  #print('fin\n',fin)
  fin['score']=fin.apply(lambda x:x['adg_score'] * x['correlation'],axis=1)
  #print('fin\n',fin)
  nume = fin['score'].sum()
  #print('nume\n',nume)
  deno = fin['correlation'].sum()
  #print('deno\n',deno)
  final_score = avg_user + (nume/deno)
  #print('final_score\n',final_score)
  score.append(final_score)
#print('list_score',score)

In [ ]:
data = pd.DataFrame({'movieId':Movies_under_consideration,'score':score})
data

movieId     score
0        3638  3.681502
1        1208  4.331430
2        4867  3.780960
3        2375  3.805372
4        2416  3.714898
...       ...       ...
1852      227  3.756604
1853     3104  3.811394
1854    63082  4.111716
1855    67255  3.791701
1856      180  3.768487

[1857 rows x 2 columns]

In [ ]:
top_5_recommendation = data.sort_values(by='score',ascending=False).head(5)
top_5_recommendation

movieId     score
476      1196  4.629604
1508      260  4.465601
114      1213  4.454659
1373     1258  4.438358
863      1201  4.402002

In [ ]:
Movie_Name = top_5_recommendation.merge(movies, how='inner', on='movieId')
Movie_Name

movieId  ...                    genres
0     1196  ...   Action|Adventure|Sci-Fi
1      260  ...   Action|Adventure|Sci-Fi
2     1213  ...               Crime|Drama
3     1258  ...                    Horror
4     1201  ...  Action|Adventure|Western

[5 rows x 4 columns]

In [ ]:
Movie_Names = Movie_Name.title.values.tolist()
print(Movie_Names)

['Star Wars: Episode V - The Empire Strikes Back (1980)', 'Star Wars: Episode IV - A New Hope (1977)', 'Goodfellas (1990)', 'Shining, The (1980)', 'Good, the Bad and the Ugly, The (Buono, il brutto, il cattivo, Il) (1966)']


#### Explain

In [ ]:
score = []
for item in Movies_under_consideration[:1]:
  print('-'*50)
  print('item: ',item)
  c = user_movie_matrix.loc[:,item] 
  print('c loc ra rating của movie đó theo toàn bộ user \n',c) # loc ra toan bo user theo colum movie dang chọn
  print(len(c))
  d = c[c.index.isin(b)]  
  print('d loc cac user id/rating average lien quan voi user da chon\n',d) # loc cac user id lien quan voi user da chon (b)
  print(len(d))
  f = d[d.notnull()] 
  print('f loai bo null user\n',f) # loai bo null user
  print(len(f))
  avg_user = Mean.loc[Mean['userId'] == userId,'avg_rating'].values[0] 
  print('avg_user\n',avg_user)#loc mean rating value cua user đang chon
  index = f.index.values.squeeze().tolist()
  print('index return index cac user id similar to user đã chọn\n',index) # return index cac user id similar to user đã chọn
  print(len(index))
  corr = user_user_sim_matrix.loc[userId,index]
  print('corr: trong danh sach similarity user x user loc ra row=user, column= list user similarity\n',corr) # trong danh sach similarity user x user loc ra row=user, column= list user similarity
  fin = pd.concat([f, corr], axis=1)
  print('fin\n',fin)
  fin.columns = ['adg_score','correlation']
  print('fin\n',fin)
  fin['score']=fin.apply(lambda x:x['adg_score'] * x['correlation'],axis=1)
  print('fin\n',fin)
  nume = fin['score'].sum()
  print('nume\n',nume)
  deno = fin['correlation'].sum()
  print('deno\n',deno)
  final_score = avg_user + (nume/deno)
  print('final_score\n',final_score)
  score.append(final_score)
print('list_score',score)

--------------------------------------------------
item:  3638
c loc ra rating của movie đó theo toàn bộ user 
 userId
316       1.893404e-16
320       4.297638e-17
359      -1.135546e-16
370       1.958963e-15
910       6.795811e-16
              ...     
137609   -1.366428e-16
137805   -1.046891e-15
138072   -1.932927e+00
138176    3.806479e-16
138200    7.808746e-16
Name: 3638, Length: 862, dtype: float64
862
d loc cac user id/rating average lien quan voi user da chon
 userId
359      -1.135546e-16
2702      8.242099e-17
2988      2.616565e-17
5611      6.647373e-16
17039    -6.767074e-17
20530    -3.724619e-17
23428     9.754102e-16
27365    -1.384174e-16
38159     0.000000e+00
40768    -5.113137e-16
41244    -3.384390e-16
42245    -9.296593e-16
43354    -5.589623e-01
45120     0.000000e+00
46645    -2.012422e-01
57937    -1.164244e+00
58265     2.139073e-01
60016     4.791753e-16
96482    -9.421488e-01
97195     5.423712e-16
101137    5.717887e-18
102118    1.843972e-02
102532    

#### End

#### Tính score từng movie này để recommend (easy way)

Score của các movie này theo 30 user similarity

In [ ]:
rating_avg

userId movieId  rating   timestamp  avg_rating  adg_rating
0        12882       1     4.0  1147195252    4.061321   -0.061321
1        12882      32     3.5  1147195307    4.061321   -0.561321
2        12882      47     5.0  1147195343    4.061321    0.938679
3        12882      50     5.0  1147185499    4.061321    0.938679
4        12882     110     4.5  1147195239    4.061321    0.438679
...        ...     ...     ...         ...         ...         ...
264500   90097   54286     5.0  1336143258    3.963415    1.036585
264501   90097   60069     4.5  1336144698    3.963415    0.536585
264502   90097   68358     5.0  1336144031    3.963415    1.036585
264503   90097   68954     5.0  1336146596    3.963415    1.036585
264504   90097   73017     4.0  1336148879    3.963415    0.036585

[264505 rows x 6 columns]

In [ ]:
print(len(Movies_under_consideration))
print(Movies_under_consideration[:10])

1857
[3638, 1208, 4867, 2375, 2416, 73, 6942, 350, 432, 3363]


In [ ]:
print(all_user_similar)

[46645, 42245, 40768, 23428, 123707, 60016, 45120, 113645, 97195, 102118, 58265, 113540, 102532, 120782, 17039, 117007, 101137, 57937, 27365, 41244, 5611, 20530, 2702, 38159, 359, 43354, 117144, 96482, 2988, 108195]


In [ ]:
# Loc dach sach cac movie rating co user similar và under consideration
rating_avg_filter=rating_avg[rating_avg.userId.isin(all_user_similar) & 
                            rating_avg.movieId.isin(list(map(str, Movies_under_consideration)))]
rating_avg_filter
# adg_rating: gia tri rating cua user tren movie do - gia tri rating trung binh cua user do tren toan bo movie
# neu adg_rating >0 : phim nay hay hon muc trung binh cua nguoi do, neu < 0 thi phim nay do hon muc trung binh cua nguoi do
# adg_rating average của các movie sẽ cho ta biết mưc độ hay dở thế nào...

userId movieId  rating   timestamp  avg_rating  adg_rating
3877    108195       5     2.0  1095774393    2.924271   -0.924271
3879    108195       7     2.0  1095774409    2.924271   -0.924271
3880    108195      10     3.5  1095772767    2.924271    0.575729
3881    108195      11     3.0  1095755740    2.924271    0.075729
3882    108195      14     2.0  1095774403    2.924271   -0.924271
...        ...     ...     ...         ...         ...         ...
260569   20530   91658     3.0  1367429314    3.439785   -0.439785
260572   20530   94959     4.0  1367427353    3.439785    0.560215
260575   20530   97304     3.5  1367427378    3.439785    0.060215
260577   20530   97938     3.0  1367427365    3.439785   -0.439785
260581   20530  109487     4.5  1423428718    3.439785    1.060215

[9967 rows x 6 columns]

In [ ]:
movie_top_10=rating_avg_filter.groupby('movieId')['adg_rating'].mean().sort_values(ascending=False)[:10]
movie_top_10

movieId
2132    1.560215
1267    1.350755
123     1.314526
930     1.298758
3526    1.234043
1131    1.234043
521     1.234043
1132    1.234043
82      1.234043
898     1.149194
Name: adg_rating, dtype: float64

In [ ]:
list(movie_top_10.index)

['2132', '1267', '123', '930', '3526', '1131', '521', '1132', '82', '898']

In [ ]:
movies_names=movies.set_index('movieId')['title']
movies_names

movieId
1                           Toy Story (1995)
2                             Jumanji (1995)
3                    Grumpier Old Men (1995)
4                   Waiting to Exhale (1995)
5         Father of the Bride Part II (1995)
                         ...                
109487                   Interstellar (2014)
111362     X-Men: Days of Future Past (2014)
111759               Edge of Tomorrow (2014)
112556                      Gone Girl (2014)
112852        Guardians of the Galaxy (2014)
Name: title, Length: 2500, dtype: object

In [ ]:
list(map(int,list(movie_top_10.index)))

[2132, 1267, 123, 930, 3526, 1131, 521, 1132, 82, 898]

In [ ]:
movies_names.loc[list(map(int,list(movie_top_10.index)))]

movieId
2132            Who's Afraid of Virginia Woolf? (1966)
1267                  Manchurian Candidate, The (1962)
123      Chungking Express (Chung Hing sam lam) (1994)
930                                   Notorious (1946)
3526                                 Parenthood (1989)
1131                           Jean de Florette (1986)
521                           Romeo Is Bleeding (1993)
1132    Manon of the Spring (Manon des sources) (1986)
82                     Antonia's Line (Antonia) (1995)
898                     Philadelphia Story, The (1940)
Name: title, dtype: object

## Item-based Collaborative Filtering

In [24]:
%matplotlib inline
import pandas as pd
import numpy as np
from ast import literal_eval
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.metrics.pairwise import linear_kernel, cosine_similarity
import warnings; warnings.simplefilter('ignore')

In [25]:
movies=pd.read_csv('/content/drive/My Drive/Data/Movie dataset/Small/movies.csv',encoding = "ISO-8859-1")
print(movies.shape)
movies.head()

(2500, 3)


movieId  ...                                       genres
0        1  ...  Adventure|Animation|Children|Comedy|Fantasy
1        2  ...                   Adventure|Children|Fantasy
2        3  ...                               Comedy|Romance
3        4  ...                         Comedy|Drama|Romance
4        5  ...                                       Comedy

[5 rows x 3 columns]

In [44]:
titles=movies[['movieId','title']]
titles=titles.set_index('movieId')['title']
titles

movieId
1                           Toy Story (1995)
2                             Jumanji (1995)
3                    Grumpier Old Men (1995)
4                   Waiting to Exhale (1995)
5         Father of the Bride Part II (1995)
                         ...                
109487                   Interstellar (2014)
111362     X-Men: Days of Future Past (2014)
111759               Edge of Tomorrow (2014)
112556                      Gone Girl (2014)
112852        Guardians of the Galaxy (2014)
Name: title, Length: 2500, dtype: object

In [27]:
ratings=pd.read_csv('/content/drive/My Drive/Data/Movie dataset/Small/ratings.csv',encoding = "ISO-8859-1")
print(ratings.shape)
ratings.head()

(264505, 4)


userId  movieId  rating   timestamp
0   12882        1     4.0  1147195252
1   12882       32     3.5  1147195307
2   12882       47     5.0  1147195343
3   12882       50     5.0  1147185499
4   12882      110     4.5  1147195239

In [28]:
print(len(ratings.userId.unique()))
print(len(ratings.movieId.unique()))

862
2500


In [29]:
Mean = ratings.groupby(by="userId",as_index=False)['rating'].mean()
Mean.rename(columns={'rating':'avg_rating'},inplace=True)
Mean

userId  avg_rating
0       316    3.329457
1       320    3.701613
2       359    3.685474
3       370    3.794404
4       910    3.898080
..      ...         ...
857  137609    3.711538
858  137805    2.779210
859  138072    2.432927
860  138176    4.107143
861  138200    3.422249

[862 rows x 2 columns]

In [30]:
rating_avg = pd.merge(ratings,Mean,on='userId')
rating_avg # rating_x: rating in each movie; rating_y: average rating all movie

userId  movieId  rating   timestamp  avg_rating
0        12882        1     4.0  1147195252    4.061321
1        12882       32     3.5  1147195307    4.061321
2        12882       47     5.0  1147195343    4.061321
3        12882       50     5.0  1147185499    4.061321
4        12882      110     4.5  1147195239    4.061321
...        ...      ...     ...         ...         ...
264500   90097    54286     5.0  1336143258    3.963415
264501   90097    60069     4.5  1336144698    3.963415
264502   90097    68358     5.0  1336144031    3.963415
264503   90097    68954     5.0  1336146596    3.963415
264504   90097    73017     4.0  1336148879    3.963415

[264505 rows x 5 columns]

In [31]:
rating_avg['adg_rating']=rating_avg['rating']-rating_avg['avg_rating']
rating_avg.head() # phim nao co rating cao hon muc trung binh cua 1 user

userId  movieId  rating   timestamp  avg_rating  adg_rating
0   12882        1     4.0  1147195252    4.061321   -0.061321
1   12882       32     3.5  1147195307    4.061321   -0.561321
2   12882       47     5.0  1147195343    4.061321    0.938679
3   12882       50     5.0  1147185499    4.061321    0.938679
4   12882      110     4.5  1147195239    4.061321    0.438679

In [32]:
user_movie_matrix = rating_avg.pivot_table(
    index='userId', 
    columns='movieId', 
    values='adg_rating',
    aggfunc='mean'
)
user_movie_matrix

movieId    1         2         3       ...    111759    112556    112852
userId                                 ...                              
316     -0.829457       NaN       NaN  ...       NaN       NaN       NaN
320           NaN       NaN       NaN  ...       NaN       NaN       NaN
359      1.314526       NaN       NaN  ...       NaN       NaN       NaN
370      0.705596  0.205596       NaN  ...  0.705596 -0.294404 -0.794404
910      1.101920  0.101920 -0.398080  ...  0.601920       NaN       NaN
...           ...       ...       ...  ...       ...       ...       ...
137609        NaN       NaN       NaN  ...       NaN       NaN       NaN
137805   1.220790 -0.779210       NaN  ...       NaN       NaN  1.220790
138072   0.567073 -1.432927 -1.432927  ...       NaN       NaN       NaN
138176        NaN       NaN       NaN  ...       NaN       NaN       NaN
138200  -0.422249       NaN       NaN  ...       NaN       NaN       NaN

[862 rows x 2500 columns]

In [33]:
# Replacing NaN by user Average (fill na by row)
user_movie_matrix = user_movie_matrix.apply(lambda row: row.fillna(row.mean()), axis=1)

In [34]:
user_movie_matrix.T

userId         316           320     ...        138176        138200
movieId                              ...                            
1       -8.294574e-01  4.297638e-17  ...  3.806479e-16 -4.222488e-01
2        1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16
3        1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16
4        1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16
5        1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16
...               ...           ...  ...           ...           ...
109487   1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16
111362   1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16
111759   1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16
112556   1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16
112852   1.893404e-16  4.297638e-17  ...  3.806479e-16  7.808746e-16

[2500 rows x 862 columns]

In [35]:
b=cosine_similarity(user_movie_matrix.T)
b

array([[ 1.        , -0.00727701, -0.03159433, ...,  0.11531286,
         0.08126606,  0.08911925],
       [-0.00727701,  1.        ,  0.14100306, ..., -0.0733531 ,
        -0.07703056, -0.10773674],
       [-0.03159433,  0.14100306,  1.        , ..., -0.03206176,
        -0.02001517, -0.05869697],
       ...,
       [ 0.11531286, -0.0733531 , -0.03206176, ...,  1.        ,
         0.14491174,  0.2992348 ],
       [ 0.08126606, -0.07703056, -0.02001517, ...,  0.14491174,
         1.        ,  0.25652769],
       [ 0.08911925, -0.10773674, -0.05869697, ...,  0.2992348 ,
         0.25652769,  1.        ]])

In [36]:
np.fill_diagonal(b, 0 )
# Chuyen toan bo hang cheo 1 thanh 0 (user trung nhau khong co y nghia o day)

In [37]:
item_item_sim_matrix = pd.DataFrame(b)
print(item_item_sim_matrix.shape)
item_item_sim_matrix.head()

(2500, 2500)


0         1         2     ...      2497      2498      2499
0  0.000000 -0.007277 -0.031594  ...  0.115313  0.081266  0.089119
1 -0.007277  0.000000  0.141003  ... -0.073353 -0.077031 -0.107737
2 -0.031594  0.141003  0.000000  ... -0.032062 -0.020015 -0.058697
3 -0.112114  0.017774  0.090117  ...  0.006900  0.000470 -0.021005
4 -0.084925  0.157326  0.168187  ... -0.138732 -0.012425 -0.082453

[5 rows x 2500 columns]

In [38]:
item_item_sim_matrix.columns = user_movie_matrix.T.index
item_item_sim_matrix['movieId'] = user_movie_matrix.T.index
item_item_sim_matrix = item_item_sim_matrix.set_index('movieId')
item_item_sim_matrix.head()

movieId    1         2         3       ...    111759    112556    112852
movieId                                ...                              
1        0.000000 -0.007277 -0.031594  ...  0.115313  0.081266  0.089119
2       -0.007277  0.000000  0.141003  ... -0.073353 -0.077031 -0.107737
3       -0.031594  0.141003  0.000000  ... -0.032062 -0.020015 -0.058697
4       -0.112114  0.017774  0.090117  ...  0.006900  0.000470 -0.021005
5       -0.084925  0.157326  0.168187  ... -0.138732 -0.012425 -0.082453

[5 rows x 2500 columns]

Making Recommendations:

In [52]:
movied_id=109487
print('movie name: ', titles.loc[movied_id])
#top_10_similar_items to this item 
n=10
top_10_similar_items = list(
    item_item_sim_matrix\
        .loc[movied_id]\
        .sort_values(ascending=False)\
        .iloc[:n]\
    .index
)

titles.loc[top_10_similar_items]

movie name:  Interstellar (2014)


movieId
106782      Wolf of Wall Street, The (2013)
96610                         Looper (2012)
112852       Guardians of the Galaxy (2014)
104841                       Gravity (2013)
3788                Blow-Up (Blowup) (1966)
68237                           Moon (2009)
2644                         Dracula (1931)
69122                  Hangover, The (2009)
2289                     Player, The (1992)
111362    X-Men: Days of Future Past (2014)
Name: title, dtype: object

In [53]:
movied_id=1
print('movie name: ', titles.loc[movied_id])
#top_10_similar_items to this item 
n=10
top_10_similar_items = list(
    item_item_sim_matrix\
        .loc[movied_id]\
        .sort_values(ascending=False)\
        .iloc[:n]\
    .index
)

titles.loc[top_10_similar_items]

movie name:  Toy Story (1995)


movieId
3114            Toy Story 2 (1999)
4886         Monsters, Inc. (2001)
78499           Toy Story 3 (2010)
6377           Finding Nemo (2003)
4306                  Shrek (2001)
8961       Incredibles, The (2004)
364          Lion King, The (1994)
919       Wizard of Oz, The (1939)
1270     Back to the Future (1985)
150               Apollo 13 (1995)
Name: title, dtype: object